### PR TITLE
Add ResLogit implementation and a notebook example

### PR DESCRIPTION
**Pull request by julesdesir**
## Description of the goal of the PR

Add ResLogit implementation and a notebook example

## Changes this PR introduces (fill it before implementation)

- Add ResLogit implementation
- Add a notebook example for ResLogit
- Fix typos in other files

## Checklist before requesting a review
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have created / updated the docstrings
- [x] I have updated the README, if relevant
- [X] I have tested my code
- [X] The CI pipeline passes
- [X] I have performed a self-review of my code

### CODE REVIEW

File: choice_learn/models/reslogit.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+import logging
+
+import tensorflow as tf
+
+import choice_learn.tf_ops as tf_ops
+from choice_learn.models.base_model import ChoiceModel
+
+
+class ResNetLayer(tf.keras.layers.Layer):
+    """The ResNet layer class."""
```
Comment (VincentAuriau) [inline choice_learn/models/reslogit.py:12]: En terme de naming, avoir NetLayer c'est un étonnant je trouve. Vu qu'il n'y a qu'une couche je parlerais plus de ResLayer et si on les additionne, ça donne un ResNet.

File: choice_learn/models/reslogit.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+
+    def __init__(self):
+        """Initialize the ResNetLayer class."""
+        super().__init__()
+
+    def get_activation_function(self, name):
+        """Get an activation function from its str name.
+
```
Comment (VincentAuriau) [inline choice_learn/models/reslogit.py:20]: https://www.tensorflow.org/api_docs/python/tf/keras/activations/get à tester, fait pê le même job

File: choice_learn/models/reslogit.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+        # Optimization parameters
+        self.label_smoothing = label_smoothing
+        self.tolerance = tolerance
+        self.lr = lr
+        self.epochs = epochs
+        self.batch_size = batch_size
+        self.logmin = logmin
+
+        if optimizer == "Adam":
```
Comment (VincentAuriau) [inline choice_learn/models/reslogit.py:188]: Is it really needed or is already from ChoiceModel ?

File: choice_learn/models/reslogit.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+        self.indexes = indexes
+        self.mnl_weights = mnl_weights
+        self.residual_weights = residual_weights
+        self._trainable_weights = weights
+        return indexes, weights
+
+    @property
+    def trainable_weights(self):
+        """Trainable weights of the model."""
+        return self._trainable_weights
```
Comment (VincentAuriau) [inline choice_learn/models/reslogit.py:335]: Si on utilise plus build() manuellement dans le instantiate, il est possible qu'il faille transformer ici en:
return self.resnet_model.trainable_variables + self.mnl_weights

File: pyproject.toml [hunk 5]
```diff
@@ -16,7 +16,7 @@ select = [
   "PTH",
   "PD",
 ] # See: https://beta.ruff.rs/docs/rules/
-ignore = ["D203", "D213", "ANN101", "ANN102", "ANN204", "ANN001", "ANN002", "ANN202", "ANN201", "ANN206", "ANN003", "PTH100", "PTH118", "PTH123"]
+ignore = ["D203", "D213", "ANN101", "ANN102", "ANN204", "ANN001", "ANN002", "ANN202", "ANN201", "ANN206", "ANN003", "PTH100", "PTH118", "PTH123","PTH113", "PTH104", "RET505"]
```
Comment (VincentAuriau) [inline pyproject.toml:19]: Discutons du pourquoi RET505

File: tests/integration_tests/models/test_reslogit.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
+"""Tests ResLogit."""
+
+import numpy as np
+
+from choice_learn.datasets import load_swissmetro
+from choice_learn.models import ResLogit, SimpleMNL
+
+dataset = load_swissmetro()
```
Comment (VincentAuriau) [inline tests/integration_tests/models/test_reslogit.py:8]: if tests are too slow, could only keep first rows or dummy dataset

File: choice_learn/models/reslogit.py [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+
+        Returns
+        -------
+        indexes : dict
+            Dictionary of the indexes of the weights created
+        weights : list of tf.Variable
+            List of the weights created coresponding to the specification
+        """
+        # Instantiate the loss function
+        self.loss = tf_ops.CustomCategoricalCrossEntropy(
```
[the remark this answers is not in the document]
Comment (julesdesir) [reply choice_learn/models/reslogit.py:220]: The ChoiceModel class doesn't have a logmin attribute.
In base_model.py:
self.loss = tf_ops.CustomCategoricalCrossEntropy(from_logits=False, label_smoothing=self.label_smoothing)
vs in reslogit.py:
self.loss = tf_ops.CustomCategoricalCrossEntropy(from_logits=False, label_smoothing=self.label_smoothing, epsilon=self.logmin)

Should I remove the logmin attribute from the ResLogit class and remove the epsilon=self.logmin argument in tf_ops.CustomCategoricalCrossEntropy()?

File: choice_learn/models/reslogit.py [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+                    residual_weights.append(layer.residual_weights)
+                else:
+                    layer.build(
+                        input_shape=(self.res_layers_width[i - 2],),
+                        layer_width=self.res_layers_width[i - 1],
+                        activation=self.activation,
+                    )
+                    residual_weights.append(layer.residual_weights)
+                output = layer(output)
+        resnet_model = tf.keras.Model(
```
[the remark this answers is not in the document]
Comment (julesdesir) [reply choice_learn/models/reslogit.py:317]: The hidden layers have an input shape depending on the width of the previous layer. I don't see how to handle the different cases without using .build for each layer separately

File: notebooks/models/reslogit.ipynb
Comment (VincentAuriau) [inline notebooks/models/reslogit.ipynb:1]: Petite typo: "Firs, we create a ChoiceDataset from the dataframe."

File: tests/integration_tests/models/test_reslogit.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+dataset = dataset[:10]  # Reduce the dataset size for faster testing
+n_items = np.shape(dataset.items_features_by_choice)[2]
+n_shared_features = np.shape(dataset.shared_features_by_choice)[2]
+n_items_features = np.shape(dataset.items_features_by_choice)[3]
+
+
+def test_reslogit_fit_with_sgd():
+    """Tests that ResLogit can fit with SGD."""
```
Comment (VincentAuriau) [inline tests/integration_tests/models/test_reslogit.py:18]: Dans un des tests avec .fit tu peux rajouter:
tf.config.run_functions_eagerly(True)

Normalement ça devrait aider pour le calcul de coverage qui marche mal avec tf.function

**Review by VincentAuriau — COMMENTED** [review]
Deux micros changements et on est bon :)

**Comment by VincentAuriau** [comment]
Great !

**Review by VincentAuriau — APPROVED** [review]
